bundles/org.openhab.core.voice/src/main/java/org/openhab/core/voice/internal/cache/AudioStreamCacheWrapper.java

```diff
@@ -0,0 +1,189 @@
+ */
+@NonNullByDefault
+public class AudioStreamCacheWrapper extends FixedLengthAudioStream {
+
+    private final Logger logger = LoggerFactory.getLogger(AudioStreamCacheWrapper.class);
+
+    private TTSResult ttsResult;
+    private int offset = 0;
+
+    // A fallback mechanism : if reading from the cache fails,
+    // It will therefore try to fallback to a direct tts stream
+    private AudioStreamSupplier fallbackDirectSupplier;
+    @Nullable
+    private AudioStream fallbackDirectAudioStream;
+
+    /***
+     * Construct a transparent AudioStream wrapper around the data from the cache.
+     *
+     * @param ttsResult The parent cached {@link TTSResult}
+     * @param supplier A fallback {@link AudioStreamSupplier}, if something goes wrong with the cache mechanism
+     */
+    public AudioStreamCacheWrapper(TTSResult ttsResult, AudioStreamSupplier supplier) {
+        super();
+        this.ttsResult = ttsResult;
+        this.fallbackDirectSupplier = supplier;
+    }
+
+    @Override
+    public AudioFormat getFormat() {
+        return ttsResult.getAudioFormat();
+    }
+
+    @Override
+    public int available() throws IOException {
+        return ttsResult.availableFrom(offset);
+    }
+
+    @Override
+    public int read() throws IOException {
+        if (fallbackDirectAudioStream == null) {
+            try {
+                byte[] bytesRead = ttsResult.read(offset, 1);
+                if (bytesRead.length == 0) {
+                    return -1;
+                } else {
+                    offset++;
+                    return bytesRead[0] & 0xff;
+                }
+            } catch (IOException e) {
+                logger.debug("Cannot read from tts cache. Will use the fallback TTS mechanism", e);
+            }
+        }
+
+        // beyond this point, we failed, so the fallback must be active :
+        enableFallback();
+        AudioStream fallbackDirectAudioStreamFinal = fallbackDirectAudioStream;
+        if (fallbackDirectAudioStreamFinal != null) {
+            return fallbackDirectAudioStreamFinal.read();
+        }
+
+        throw new IOException("Neither TTS cache nor TTS fallback method succeed");
+    }
+
+    @Override
+    public int read(byte @Nullable [] b, int off, int len) throws IOException {
+        if (fallbackDirectAudioStream == null) {
+            if (b == null) {
+                throw new IOException("Array to write is null");
+            }
+            Objects.checkFromIndexSize(off, len, b.length);
+
+            if (len == 0) {
+                return 0;
+            }
+
+            try {
+                byte[] bytesRead = ttsResult.read(offset, len);
+                offset += bytesRead.length;
+                if (bytesRead.length == 0) {
+                    return -1;
+                }
+                int i = 0;
+                for (; i < len && i < bytesRead.length; i++) {
+                    b[off + i] = bytesRead[i];
```

Comment:
There is only one TTSResult, but many AudioStream can be instanciated from it (even concurrently)
If the TTSResult implements InputStream, it wouldn't be able to be reused (because once closed it is useless)
The TTSResults are made to be stored as LRU entries, and reused at will by creating several AudioStream from it.

(But it's also equally possible that christmas put a toll on my mind and that I don't understand your proposition 😅 ?)